content/posts/auto-sync/index.md

```diff
@@ -0,0 +1,448 @@
+
+void decodeOperandA(MCInst *MI, unsigned OpNum, unsigned Val) {
+-	if (MCInst_isPredicable(MI)) {
++	if (MI.isPredicable()) {
+  	MCOperand_CreateImm0(MI, Val + 1);
+  }
+	MCOperand_CreateImm0(MI, Val);
+}
+═════════════════════════════════════════════════════════════════════════════════════════════════════════
+
+Choice: O, o, n, s (none) , e, p, q, ? > ?
+O		- Accept ALL old diffs
+o		- Accept old diff
+n		- Accept new diff
+e		- Edit diff (not yet implemented)
+s (none) 	- Select saved choice
+p		- Ignore and go to previous diff
+q		- Quit (previous selections will be saved)
+?		- Show this help
+```
+
+They can accept the version from the old file or accept the version from the new file.
+The version from the new file would not compile, but it can be fixes by hand later.
+
+In most cases though, the old version is the correct one, because it was fixed by someone before.
+
+The diffing happens for each translated function which doesn't match the old code.
+Of course, you can not just diff functions but any nodes in the AST of a file.
+And for convenience the choices are saved as well. So if the update is run again and nothing changed, the user doesn't have to redo previous decisions.
+
+This diffing step saves a lot of time.
+We could require from the user to figure this problem out on their own. Searching for a proper diffing tool and going over each file.
+But this assumes that there is tool for this task on every OS and that it is convenient to use. None of which is guaranteed.
+
+## Adding new architectures
+
+The whole procedure works pretty much the same if we want to add new architecture.
+
+Generally though it gives us a standardized way of doing it. And if you know one architecture module in Capstone, you know them all.
+The only thing one needs, is the LLVM support of the architecture or a fork which has it implemented.
+
+In fact, we added two niche architectures this way. TriCore was only implemented in a fork and never upstreamed.
+And the Alpha architecture support was dropped in an earlier LLVM version.
+
+## Wrap up
+
+If one looks at all of it, it is still a rather complicated number of steps to take. But the result is worth it.
+The amount of time someone has to spend for updating an architecture module in Capstone went down from "no one did it" to roughly 4-19 hours.
+To update the ARM architecture module to LLVM 16 for example, the times were:
+
+1. Rebasing patched backends to new LLVM release = ~1-3h
+2. Running the update scripts and diffing = 5min - 1h
+3. Fixing rest of build errors by hand = ~30min - 5h
+4. Handle new operands on the CS side (filling the detail info and tests) - 3-10h
+
+(Please be aware though, that the time estimates from above don't include the "read into" time someone has to spend.)
+
+The lower estimates are for small changes on the LLVM side, the upper for many changes (spanning multiple LLVM releases).
```

Comment:
Please add also a paragraph about fixes you sent back to the LLVM, with links, and possible ideas to fix more.

Reply:
Added the LLVM fixes to the future task list. Because I think a complete list is not really within the scope of this blog post.